File: ios/MediaPipe/ObjectAnalyzerBridge.swift
import Foundation
import MediaPipeTasksVision
import UIKit

/// Wraps MediaPipe Tasks [ObjectDetector] in IMAGE running mode.
///
/// The `.tflite` model file must be bundled into the Runner target's resources.
final class ObjectAnalyzerBridge {
    private let detector: ObjectDetector

    init() throws {
        let modelPath = try Self.requireModelPath(name: "efficientdet_lite0", ext: "tflite")
        let options = ObjectDetectorOptions()
        options.baseOptions.modelAssetPath = modelPath
        options.runningMode = .image
        options.maxResults = 5
        options.scoreThreshold = 0.5
        detector = try ObjectDetector(options: options)
    }

    func detect(frame: FrameArgs) throws -> [[String: Any]] {
        guard let uiImage = FrameDecoder.decodeUprightImage(frame: frame) else {
            return []
        }
        let mpImage = try MPImage(uiImage: uiImage)
        let result = try detector.detect(image: mpImage)

        var out: [[String: Any]] = []
        for detection in result.detections {
            let category = detection.categories.first
            let box = detection.boundingBox
            out.append([
                "label": category?.categoryName ?? "unknown",
                "confidence": Double(category?.score ?? 0),
                "bbox": [
                    "left": Double(box.origin.x),
                    "top": Double(box.origin.y),
                    "width": Double(box.size.width),
                    "height": Double(box.size.height),
                ],
            ])
        }
        return out
    }

    private static func requireModelPath(name: String, ext: String) throws -> String {
        guard let path = Bundle.main.path(forResource: name, ofType: ext) else {
            throw NSError(
                domain: "MediaPipePlugin",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Missing bundled model: \(name).\(ext)"],
            )
        }
        return path
    }
}

File: android/app/src/main/kotlin/com/example/face_liveness_demo/mediapipe/ObjectAnalyzerBridge.kt
package com.example.face_liveness_demo.mediapipe

import android.content.Context
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.objectdetector.ObjectDetector
import com.google.mediapipe.tasks.vision.objectdetector.ObjectDetectorResult

/**
 * Wraps MediaPipe Tasks [ObjectDetector].
 *
 * Model file must be bundled at `android/app/src/main/assets/efficientdet_lite0.tflite`.
 */
class ObjectAnalyzerBridge(context: Context) {
    private val detector: ObjectDetector

    init {
        val options = ObjectDetector.ObjectDetectorOptions.builder()
            .setBaseOptions(
                BaseOptions.builder()
                    .setModelAssetPath(MODEL_ASSET_PATH)
                    .build(),
            )
            .setRunningMode(RunningMode.IMAGE)
            .setMaxResults(MAX_RESULTS)
            .setScoreThreshold(SCORE_THRESHOLD)
            .build()
        detector = ObjectDetector.createFromOptions(context, options)
    }

    fun detect(frame: FrameArgs): List<Map<String, Any>> {
        val bitmap = FrameDecoder.decodeUprightBitmap(frame) ?: return emptyList()
        val mpImage = BitmapImageBuilder(bitmap).build()
        val result: ObjectDetectorResult = detector.detect(mpImage)

        val out = mutableListOf<Map<String, Any>>()
        for (detection in result.detections()) {
            val box = detection.boundingBox() ?: continue
            val category = detection.categories().firstOrNull()
            out += mapOf(
                "label" to (category?.categoryName() ?: "unknown"),
                "confidence" to (category?.score()?.toDouble() ?: 0.0),
                "bbox" to mapOf(
                    "left" to box.left.toDouble(),
                    "top" to box.top.toDouble(),
                    "width" to box.width().toDouble(),
                    "height" to box.height().toDouble(),
                ),
            )
        }
        return out
    }

    fun close() {
        try {
            detector.close()
        } catch (_: Throwable) {
        }
    }

    companion object {
        private const val MODEL_ASSET_PATH = "efficientdet_lite0.tflite"
        private const val MAX_RESULTS = 5
        private const val SCORE_THRESHOLD = 0.5f
    }
}
